### ISSUE TITLE
Error selecting records with shift+arrow keys

### ISSUE THREAD

**Issue by kravlost** [HeidiSQL/HeidiSQL#1451]
**Preconditions**

* HeidiSQL version: 11.3.0.6295 64-bit
* Database type and version: Postgres 12.dll
* OS: Windows 10

**Describe the bug**

Shift+arrow keys going past the end of the data and then back causes a fault.

**To Reproduce**

Example:
1. Go to the data view
2. Go near the bottom of the data.
3. Hold down shift. Press down arrow to select items.
4. Go one past the final record. The selected rows are unselected, and a final, empty row is selected.
5. Press up arrow to go back onto the final record (still with shift held down)
6. An error box appears

**Screenshots or Crash reports**

operating system   : Windows 10 x64 build 19043
system language    : English
program up time    : 7 minutes 4 seconds
process id         : $239c
allocated memory   : 79.10 MB
largest free block : 131021.95 GB
executable         : heidisql.exe
exec. date/time    : 2021-05-30 07:40
version            : 11.3.0.6295
compiled with      : Delphi 10.4 Sydney
madExcept version  : 5.1.0
callstack crc      : $68f96c16, $6d545193, $6d545193
exception number   : 5
exception class    : EAccessViolation
exception message  : Access violation at address 000000000095F45E in module 'heidisql.exe'. Read of address FFFFFFFFFFFFFFFF.

thread $188:
0095f45e heidisql.exe VirtualTrees   13345   +2 TBaseVirtualTree.GetFullyVisible
00985ecd heidisql.exe VirtualTrees   29773   +5 TBaseVirtualTree.GetVisibleParent
00985189 heidisql.exe VirtualTrees   29200  +11 TBaseVirtualTree.GetPreviousVisible
0097d3db heidisql.exe VirtualTrees   25346   +8 TBaseVirtualTree.ToggleSelection
009687b6 heidisql.exe VirtualTrees   17183 +423 TBaseVirtualTree.WMKeyDown
0040f072 heidisql.exe System                    TObject.Dispatch
006cd90d heidisql.exe Vcl.Controls              TControl.WndProc
006d4b6b heidisql.exe Vcl.Controls              TWinControl.WndProc
0097e1b3 heidisql.exe VirtualTrees   25808  +32 TBaseVirtualTree.WndProc
006d3cea heidisql.exe Vcl.Controls              TWinControl.MainWndProc
00569013 heidisql.exe System.Classes            StdWndProc
7ff8c89d USER32.dll                             DispatchMessageW
0088c44e heidisql.exe Vcl.Forms                 TApplication.ProcessMessage
0088c4c3 heidisql.exe Vcl.Forms                 TApplication.HandleMessage
0088c911 heidisql.exe Vcl.Forms                 TApplication.Run
01066588 heidisql.exe heidisql         104  +41 initialization
7ff8c8dc KERNEL32.DLL                           BaseThreadInitThunk
7ff8c92e ntdll.dll                              RtlUserThreadStart

**Comment by kravlost** [comment]
@ansgarbecker Has this been fixed yet?

**Comment by ansgarbecker** [comment]
No, and everything I tried yet fails with the same exception callstack.
The tree somehow wants to toggle selected nodes if you use the shift key, and HeidiSQL removes the cancelled node too early. I moved the deletion from `OnFocusChanging` to `OnFocusChanged`, but for some reason that is still too early.

**Comment by kravlost** [comment]
Any progress with this?

**Comment by ansgarbecker** [comment]
Still crashing in v12.15, only with a slightly different callstack:

```
exception class    : EAccessViolation
exception message  : Access violation at address 0000000000B61EE8 in module 'heidisql.exe' (offset 5C1EE8). Read of address 0000000A000A0089.

main thread ($2ecc):
00b61ee8 heidisql.exe VirtualTrees.BaseTree 19485   +9 TBaseVirtualTree.GetTopInvisibleParent
00b60d82 heidisql.exe VirtualTrees.BaseTree 18849  +10 TBaseVirtualTree.GetPreviousVisible
00b58e6b heidisql.exe VirtualTrees.BaseTree 15021   +8 TBaseVirtualTree.ToggleSelection
00b45106 heidisql.exe VirtualTrees.BaseTree  7147 +423 TBaseVirtualTree.WMKeyDown
005af6f1 heidisql.exe System                           TObject.Dispatch
0086eca7 heidisql.exe Vcl.Controls                     TControl.WndProc
008769bb heidisql.exe Vcl.Controls                     TWinControl.WndProc
00b59a4e heidisql.exe VirtualTrees.BaseTree 15405  +32 TBaseVirtualTree.WndProc
00875a8a heidisql.exe Vcl.Controls                     TWinControl.MainWndProc
00782793 heidisql.exe System.Classes                   StdWndProc
7ffcbecf USER32.dll                                    DispatchMessageW
00a1c1f0 heidisql.exe Vcl.Forms                        TApplication.ProcessMessage
00a1c263 heidisql.exe Vcl.Forms                        TApplication.HandleMessage
00a1c6ba heidisql.exe Vcl.Forms                        TApplication.Run
013fccb9 heidisql.exe heidisql                118  +44 initialization
```

The v13 preview release and the Linux + macOS releases do not have that crash. Having shift pressed there while on the last row blocks going further down for creating a new row. This blocking is maybe a good bugfix idea for the classic Windows version.

**Comment by ansgarbecker** [comment]
Huh, I already fixed that in the Linux version some weeks ago for #2284 - without knowing that modification would have also fixed this issue in the Windows version.
